### PR TITLE
use cmake to obtain correct soname

### LINKED ISSUE #69: SONAMEs follow different scheme than upstream (e.g. `so.13.5.2` vs `so.18`)
### Solution to issue cannot be found in the documentation.

- [X] I checked the documentation.

### Issue

`libarchive`, as packaged on conda-forge, follows a different SONAME scheme than the one [suggested in upstream](https://github.com/libarchive/libarchive/blob/464f9d9f1a5b3f51845e5f8662a37f48a7dfd996/CMakeLists.txt#L71-L81):

```cmake
# INTERFACE_VERSION increments with every release
# libarchive 2.7 == interface version 9 = 2 + 7
# libarchive 2.8 == interface version 10 = 2 + 8
# libarchive 2.9 == interface version 11 = 2 + 9
# libarchive 3.0 == interface version 12
# libarchive 3.1 == interface version 13
math(EXPR INTERFACE_VERSION  "13 + ${_minor}")

# Set SOVERSION == Interface version
# ?? Should there be more here ??
SET(SOVERSION "${INTERFACE_VERSION}")
```

For 3.5.2, this means `13+5=18`, so `libarchive.so.18`. However, the contents of the tarballs list `libarchive.so.13.5.2` (?!) and, most importantly, it includes a symlink from `libarchive.so.13 -> libarchive.so.13.5.2`, which would make this version collide with a (correctly SONAMEd) libarchive 3.1.

Compare this to defaults:

# conda-forge

```
libarchive-3.5.2-hb890918_2.tar.bz2
-rw-rw-r--  0 1001   1001  1818396 May 18 09:00 lib/libarchive.a
-rw-rw-r--  0 1001   1001      871 May 18 09:00 lib/pkgconfig/libarchive.pc
-rwxrwxr-x  0 1001   1001   921896 May 18 09:00 lib/libarchive.so.13.5.2
lrwxrwxrwx  0 1001   1001        0 May 18 09:00 lib/libarchive.so -> libarchive.so.13.5.2
lrwxrwxrwx  0 1001   1001        0 May 18 09:00 lib/libarchive.so.13 -> libarchive.so.13.5.2
libarchive-3.5.2-hccf745f_0.tar.bz2
lrwxrwxrwx  0 1001   1001        0 Aug 27  2021 lib/libarchive.so -> libarchive.so.13.5.2
-rw-rw-r--  0 1001   1001      871 Aug 27  2021 lib/pkgconfig/libarchive.pc
-rwxrwxr-x  0 1001   1001   916664 Aug 27  2021 lib/libarchive.so.13.5.2
lrwxrwxrwx  0 1001   1001        0 Aug 27  2021 lib/libarchive.so.13 -> libarchive.so.13.5.2
-rw-rw-r--  0 1001   1001  1808436 Aug 27  2021 lib/libarchive.a
libarchive-3.5.2-hccf745f_1.tar.bz2
lrwxrwxrwx  0 1001   1001        0 Sep 25  2021 lib/libarchive.so.13 -> libarchive.so.13.5.2
-rw-rw-r--  0 1001   1001  1808436 Sep 25  2021 lib/libarchive.a
lrwxrwxrwx  0 1001   1001        0 Sep 25  2021 lib/libarchive.so -> libarchive.so.13.5.2
-rw-rw-r--  0 1001   1001      871 Sep 25  2021 lib/pkgconfig/libarchive.pc
-rwxrwxr-x  0 1001   1001   916664 Sep 25  2021 lib/libarchive.so.13.5.2
```

# defaults

```
libarchive-3.5.2-h5de8990_0.tar.bz2
-rw-rw-r--  0 0      0         930 Jun  3 13:21 lib/pkgconfig/libarchive.pc
lrwxrwxrwx  0 0      0           0 Jun  3 13:22 lib/libarchive.so -> libarchive.so.18
-rw-rw-r--  0 0      0     1813266 Jun  3 13:22 lib/libarchive.a
-rwxrwxr-x  0 0      0      912576 Jun  3 13:22 lib/libarchive.so.18
```

---

Looks like defaults relies on CMake while conda-forge uses autotools, which might explain the difference. I think the bug is that what should be an arithmetical addition (13 + 5) is being treated as a string concatenation (13 + "5.2"). 

Fixing this would involve rebuilding downstream packages I guess.

### Installed packages

```shell
N/A
```

### Environment info

```shell
N/A
```

### PR DESCRIPTION
**Pull request by jaimergp**
Checklist
* [X] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

See #69 for more details

### CODE REVIEW

**Comment by jaimergp** [comment]
@conda-forge-admin, please rerender

File: recipe/build.sh
Comment (ocefpaf) [inline recipe/build.sh]: I'm assuming we won't need any repo patching or rebuilds, right? This ensure the old packages built with this will work.

I'm just concern of the sustainability of this `else` but I guess this is better than causing more churn in the ecosystem.

**Review by ocefpaf — APPROVED** [review]

File: recipe/build.sh
Comment (jaimergp) [reply recipe/build.sh]: Existing packages should work on Linux, but there's no workaround on macOS :(

See [this comment](https://github.com/conda-forge/libarchive-feedstock/issues/69#issuecomment-1152256077)

Comment (jaimergp) [reply recipe/build.sh]: Also I'd say we should only add the symlinks in this transition PR. Next release (e.g. 3.6) should not add the buggy sonames and just stick to upstream conventions (should be so.19 by then).

Comment (ocefpaf) [reply recipe/build.sh]: I'm OK with that. Let's wait for some else to comment before merging just so we have third opinion.

@jakirkham you are the only other maintainer here BTW :grimacing:

Comment (jaimergp) [reply recipe/build.sh]: If possible I'd like to wait til next core call to discuss it with the team. We might need to rebuild libmamba on osx among others.

Comment (jakirkham) [reply recipe/build.sh]: Yeah not sure about this `else` either

Also feel a little weird about mucking with the SONAME like this generally. Have we tried raising an issue with upstream?

Comment (jaimergp) [reply recipe/build.sh]: In my opinion the autoconf workflow has a bug. The expected so names are single integers according to their own docs:

* [Configure.ac](https://github.com/libarchive/libarchive/blob/464f9d9f1a5b3f51845e5f8662a37f48a7dfd996/configure.ac#L35-L47)
* [CMakeLists.txt](https://github.com/libarchive/libarchive/blob/464f9d9f1a5b3f51845e5f8662a37f48a7dfd996/CMakeLists.txt#L71-L77)

It's unclear to me why we ended up with these weird `13.5.2` so names. Looks like `13` + the minor.patch version, string-concatenated, instead of arithmetically summed. CMake does produce `18`.

There have been some similar issues in other distros, in which they couldn't agree on the soname version, but all of them use single integers:

* https://github.com/AppImage/AppImageKit/issues/343, https://github.com/AppImage/AppImageKit/issues/476
* https://github.com/libarchive/libarchive/issues/817, https://github.com/libarchive/libarchive/issues/749

**Comment by jaimergp** [comment]
@ocefpaf, this is probably going to become a bigger issue when `libmamba` becomes default, and I am seeing reports of the defaults<->conda-forge incompatibility in more and more places:

- https://github.com/SciTools/workflows/pull/3
- https://github.com/conda/conda/issues/13041
- https://github.com/conda/conda-libmamba-solver/issues/244
- https://github.com/mamba-org/mamba/issues/1826 (this has a handful!)
- ... and probably more

As a result I'd like to move forward with this PR. I've added symlinks to ensure backwards compatibility with the previous builds and a "release bomb" for 3.8 so we remove them. We'll also need to backport this to the 3.6 and 3.5 branches.

If it doesn't work, well, I'll mark them as broken and go back to the drawing board 😬

**Comment by JeanChristopheMorinPerso** [comment]
Hi @jaimergp I did a quick check with defaults and I see that we build with `-DENABLE_OPENSSL=TRUE` on unix systems (except for osx-64 where we don't build with OpenSSL, but I don't know why). Is it something that should be replicated here too? See https://github.com/AnacondaRecipes/libarchive-feedstock/blob/master/recipe/build.sh#L12-L16.

File: recipe/build_libarchive.sh
Comment (jaimergp) [inline recipe/build_libarchive.sh:33]:
```suggestion
    -DENABLE_CNG=FALSE                 \
    -DENABLE_OPENSSL=TRUE              \
```

**Comment by jaimergp** [comment]
@JeanChristopheMorinPerso - True, good catch! I missed that line in `configure`. Added it now. Thanks!

**Comment by JeanChristopheMorinPerso** [comment]
Np. But you know what, it's actually the default. I didn't realize it is the default. Anyway, it doesn't hurt to be explicit! Thanks for making the modification!

**Comment by jaimergp** [comment]
Hm, @isuruf pointed me to this [list of items to check](https://conda-forge.org/docs/maintainer/knowledge_base.html#moving-from-an-autotools-build-to-a-cmake-build) when making this precise change (autotools -> CMake), and it might be trickier than expected (or not possible at all). I'll report later.

**Comment by jaimergp** [comment]
Closing here as we won't be able to fix it this way. Check the issue for more details on the strategy going forward.

**Pull request closed without merging**
